### PR TITLE
Replace sympy-based parser in circuit_from_quirk_url with ported parser from Quirk

### PR DESCRIPTION
**Pull request by Strilanc**
Fixes https://github.com/quantumlib/Cirq/issues/2259

### CODE REVIEW

File: cirq/contrib/quirk/cells/parse.py [hunk 1]
```diff
@@ -12,121 +12,376 @@
...
+        self.func = func
+        self.weight = weight
+
+
+_HangingToken = Union[_ResolvedToken, str, _CustomQuirkOperationToken]
+
+
+def _translate_token(token_id: str,
+                     token_map: Mapping[str, _HangingToken]) -> _HangingToken:
+    if re.match(r'[0-9]+(\.[0-9]+)?', token_id):
```
Comment (viathor) [inline cirq/contrib/quirk/cells/parse.py]: Note that `re.match` matches the regex *at the beginning* of text, i.e. `re.match('a', 'ab')` is considered a match (see also [documentation](https://docs.python.org/3/library/re.html)). Therefore, for defense-in-depth I'd append `$` to the regex to force exact match:

```python
if re.match(r'[0-9]+(\.[0-9]+)?$', token_id):
```

This way someone reading the code doesn't have to wonder whether `float(token_id)` raises `ValueError` or not.

File: cirq/contrib/quirk/cells/parse.py [hunk 2]
```diff
@@ -12,121 +12,376 @@
...
+    token_map = {**PARSE_COMPLEX_TOKEN_MAP_RAD, 't': sympy.Symbol('t')}
+    result = _parse_formula_using_token_map(formula, token_map)
+
+    if isinstance(result, sympy.Basic):
+        if result.free_symbols:
+            return result
+        result = complex(result)
+
+    if isinstance(result, complex):
+        if abs(np.imag(result)) > 1e-8:
```
Comment (viathor) [inline cirq/contrib/quirk/cells/parse.py]: nit: pass in tolerance for consistency with rest of cirq

Comment (Strilanc) [reply cirq/contrib/quirk/cells/parse.py]: Acknowledged, but that would just be a "be inconsistent with Quirk" parameter.

File: cirq/contrib/quirk/cells/parse.py (hunk 1 above)
Comment (Strilanc) [reply cirq/contrib/quirk/cells/parse.py]: Acknowledged, but I'd prefer to keep it closer to the Quirk implementation.

File: cirq/contrib/quirk/cells/parse.py [hunk 3]
```diff
@@ -12,121 +12,375 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import re
-from typing import Any, Union
+from typing import (List, TypeVar, Callable, Any, Iterator, Iterable, Union,
+                    Dict, Optional, cast, SupportsFloat, Mapping)
```
[the remark this answers is not in the document]
Comment (Strilanc) [reply cirq/contrib/quirk/cells/parse.py]: Done.

I think I'm going to start pushing back on these unless we write a tool to do it for us. I don't ever read the typing import lists in a way where it matters if they are alphabetized.

**Pull request merged**